### PR TITLE
Use(client go/log)

### LINKED ISSUE #11568: virtctl: use client-go/log
**Is your feature request related to a problem? Please describe**:
A clear and concise description of what the problem is.

We have a mix of components in virtctl that are using either `golang/glog` or `kubevirt.io/client-go/log`. 
For consistency, we should use the same logging module and I'd recommend our `client-go/log` as, if we need, we can change it latter to affect all components that use it.

I consider this a sub-task of https://github.com/kubevirt/kubevirt/issues/5932

**Describe the solution you'd like**:
- All `virtctl` modules should be using `client-go/log`
- They should all call `InitializeLogging` as well, to provide a reasonable component name.

### PR DESCRIPTION
**Pull request by Zaher1307**
### What this PR does
Before this PR:
We have a mix of components in virtctl that are using either `golang/glog` or `kubevirt.io/client-go/log`.
After this PR:
We have a all components in virtctl that are using `kubevirt.io/client-go/log`.

Fixes #11568

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered: https://kubernetes.slack.com/archives/C0163DT0R8X/p1711103910753259?thread_ts=1710776448.848269&cid=C0163DT0R8X

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: pkg/virtctl/portforward/portforwarder.go [hunk 1]
```diff
@@ -5,7 +5,7 @@ import (
 	"net"
 	"strings"
 
-	"github.com/golang/glog"
+	"kubevirt.io/client-go/log"
```
Comment (0xFelix) [inline pkg/virtctl/portforward/portforwarder.go:9]: nit: Remove newline

File: pkg/virtctl/ssh/native.go [hunk 2]
```diff
@@ -8,11 +8,11 @@ import (
 	"net"
 	"os"
 
-	"github.com/golang/glog"
 	"github.com/spf13/pflag"
 	"golang.org/x/crypto/ssh"
 	"golang.org/x/crypto/ssh/agent"
 	"golang.org/x/term"
+	"kubevirt.io/client-go/log"
```
Comment (0xFelix) [inline pkg/virtctl/ssh/native.go:15]: nit: move to L20

File: pkg/virtctl/portforward/portforward.go [hunk 3]
```diff
@@ -26,9 +26,9 @@ import (
 	"os"
 	"os/signal"
 
-	"github.com/golang/glog"
 	"github.com/spf13/cobra"
 	"k8s.io/client-go/tools/clientcmd"
+	"kubevirt.io/client-go/log"
```
Comment (0xFelix) [inline pkg/virtctl/portforward/portforward.go:31]: nit: Move down one line

File: pkg/virtctl/ssh/ssh.go [hunk 4]
```diff
@@ -24,9 +24,9 @@ import (
 	"os"
 	"path/filepath"
 
-	"github.com/golang/glog"
 	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
+	"kubevirt.io/client-go/log"
```
Comment (0xFelix) [inline pkg/virtctl/ssh/ssh.go:29]: nit: Move between k8s and local import (L32)

File: pkg/virtctl/ssh/wrapped.go [hunk 5]
```diff
@@ -2,12 +2,11 @@ package ssh
 
 import (
 	"fmt"
+	"kubevirt.io/client-go/log"
```
Comment (0xFelix) [inline pkg/virtctl/ssh/wrapped.go:5]: nit: Move to L10

File: pkg/virtctl/vnc/vnc.go [hunk 6]
```diff
@@ -32,9 +32,9 @@ import (
 	"runtime"
 	"time"
 
-	"github.com/golang/glog"
 	"github.com/spf13/cobra"
 	"k8s.io/client-go/tools/clientcmd"
+	"kubevirt.io/client-go/log"
 
 	"kubevirt.io/client-go/kubecli"
```
Comment (0xFelix) [inline pkg/virtctl/vnc/vnc.go:39]: Same, move down one line

File: pkg/virtctl/vnc/vnc.go [hunk 7]
```diff
@@ -280,29 +280,25 @@ func checkAndRunVNCViewer(doneChan chan struct{}, viewResChan chan error, port i
 	}
 
 	if vncBin == "" {
-		glog.Errorf("No supported VNC app found in %s", osType)
+		log.Log.Errorf("No supported VNC app found in %s", osType)
 		err = fmt.Errorf("No supported VNC app found in %s", osType)
 	} else {
-		if glog.V(4) {
```
Comment (0xFelix) [inline pkg/virtctl/vnc/vnc.go:286]: Keep `V(4)`

File: pkg/virtctl/vnc/vnc.go [hunk 8]
```diff
@@ -280,29 +280,25 @@ func checkAndRunVNCViewer(doneChan chan struct{}, viewResChan chan error, port i
...
-			glog.V(2).Infof("%v output: %v", vncBin, string(output))
+			log.Log.V(2).Infof("%v output: %v", vncBin, string(output))
 		}
 	}
 	viewResChan <- err
 }
 
 func tigerVncArgs(port int) (args []string) {
 	args = append(args, fmt.Sprintf(listenAddressFmt, port))
-	if glog.V(4) {
```
Comment (0xFelix) [inline pkg/virtctl/vnc/vnc.go:303]: Same

File: pkg/virtctl/vnc/vnc.go [hunk 9]
```diff
@@ -316,17 +312,13 @@ func realVncArgs(port int) (args []string) {
 	args = append(args, "-WarnUnencrypted=0")
 	args = append(args, "-Shared=0")
 	args = append(args, "-ShareFiles=0")
-	if glog.V(4) {
```
Comment (0xFelix) [inline pkg/virtctl/vnc/vnc.go:319]: Same

File: pkg/virtctl/vnc/vnc.go [hunk 10]
```diff
@@ -316,17 +312,13 @@ func realVncArgs(port int) (args []string) {
...
-		args = append(args, "-log=*:stderr:100")
-	}
+	args = append(args, "-log=*:stderr:100")
 	return
 }
 
 func remoteViewerArgs(port int) (args []string) {
 	args = append(args, fmt.Sprintf("vnc://127.0.0.1:%d", port))
-	if glog.V(4) {
```
Comment (0xFelix) [inline pkg/virtctl/vnc/vnc.go:327]: Same

**Review by 0xFelix — COMMENTED** [review]
Thanks! Looks mostly good, added some comments.

What about `AddGlogFlags`? Does it still apply?

https://github.com/kubevirt/kubevirt/blob/66a2692a4e105830cebe061464fd4282ec96655c/pkg/virtctl/log.go#L32-L40

**Comment by victortoso** [comment]
It would be lovely if you can take the opportunity and call [InitializeLogging](https://github.com/kubevirt/kubevirt/blob/main/staging/src/kubevirt.io/client-go/log/log.go#L86) too. I don't see any reason not to initialize it at least.

**Comment by alicefr** [comment]
@Zaher1307 you need to squash the merge commit. Additionally, if you enable the go formatter in your text editor it should fix all the position of the go imports. Otherwise, try to run gofmt against the files

File: pkg/virtctl/vnc/vnc.go (hunk 7 above)
Comment (Zaher1307) [reply pkg/virtctl/vnc/vnc.go:286]: client-go/log package doesn't provide any public attribute or method to figure out the current verbosity level

**Comment by Zaher1307** [comment]
> @Zaher1307 you need to squash the merge commit. Additionally, if you enable the go formatter in your text editor it should fix all the position of the go imports. Otherwise, try to run gofmt against the files

I have already formatter in my text editor (neovim) and this formatter doesn't change anything.
running `go fmt ./...` in the pkg/virtctl directory doesn't change any thing.
running `gofumpt` (a strict go formatter) doesn't change anything.
so do you have any formatting rules to follow?

File: pkg/virtctl/portforward/portforwarder.go (hunk 1 above)
Comment (Zaher1307) [reply pkg/virtctl/portforward/portforwarder.go:9]: `go fmt` doesn't change anything.
do you have any formatting rules to follow?

**Comment by Zaher1307** [comment]
> Thanks! Looks mostly good, added some comments.
> 
> What about `AddGlogFlags`? Does it still apply?
> 
> https://github.com/kubevirt/kubevirt/blob/66a2692a4e105830cebe061464fd4282ec96655c/pkg/virtctl/log.go#L32-L40

I think there is no need now for this function because glog already has been remove from kubevirtctl

**Comment by Zaher1307** [comment]
> It would be lovely if you can take the opportunity and call [InitializeLogging](https://github.com/kubevirt/kubevirt/blob/main/staging/src/kubevirt.io/client-go/log/log.go#L86) too. I don't see any reason not to initialize it at least.

There is initialization in virtctl/root.go in `Execute()` function, I think it may be useful if I called InitializeLogging in the init function of the `virtctl` package with `os.Args[0]` as program name

**Comment by victortoso** [comment]
> There is initialization in virtctl/root.go in `Execute()` function, I think it may be useful if I called InitializeLogging in the init function of the `virtctl` package with `os.Args[0]` as program name

I think it need to be per module? So, on `root.go` logs, it'll appear the component just right but on `usbredir` it'll be empty.

File: pkg/virtctl/vnc/vnc.go (hunk 7 above)
Comment (victortoso) [reply pkg/virtctl/vnc/vnc.go:286]: Should be fine to introduce one method that does return the current set filtered LogLevel. Please make this addition its own commit.

**Comment by 0xFelix** [comment]
> > Thanks! Looks mostly good, added some comments.
> > What about `AddGlogFlags`? Does it still apply?
> > https://github.com/kubevirt/kubevirt/blob/66a2692a4e105830cebe061464fd4282ec96655c/pkg/virtctl/log.go#L32-L40
> 
> I think there is no need now for this function because glog already has been remove from kubevirtctl

If there is no need for it you can drop it in an additional commit.

**Comment by 0xFelix** [comment]
> > @Zaher1307 you need to squash the merge commit. Additionally, if you enable the go formatter in your text editor it should fix all the position of the go imports. Otherwise, try to run gofmt against the files
> 
> I have already formatter in my text editor (neovim) and this formatter doesn't change anything. running `go fmt ./...` in the pkg/virtctl directory doesn't change any thing. running `gofumpt` (a strict go formatter) doesn't change anything. so do you have any formatting rules to follow?

You can run `make format` to run formatting with the rules defined in the project.

**Comment by 0xFelix** [comment]
> > There is initialization in virtctl/root.go in `Execute()` function, I think it may be useful if I called InitializeLogging in the init function of the `virtctl` package with `os.Args[0]` as program name
> 
> I think it need to be per module? So, on `root.go` logs, it'll appear the component just right but on `usbredir` it'll be empty.

Agree with Victor, can you make sure to initialize the logger in the modules it is used in?

File: pkg/virtctl/vnc/vnc.go (hunk 7 above)
Comment (victortoso) [reply pkg/virtctl/vnc/vnc.go:286]: Here it goes: https://github.com/kubevirt/kubevirt/pull/11644
(I was actually in need of such a function too!)

**Comment by Zaher1307** [comment]
> > > There is initialization in virtctl/root.go in `Execute()` function, I think it may be useful if I called InitializeLogging in the init function of the `virtctl` package with `os.Args[0]` as program name
> > 
> > 
> > I think it need to be per module? So, on `root.go` logs, it'll appear the component just right but on `usbredir` it'll be empty.
> 
> Agree with Victor, can you make sure to initialize the logger in the modules it is used in?

Can I add an init() function for every package in virtctl or you don't prefere using init functions?

File: pkg/virtctl/vnc/vnc.go (hunk 7 above)
Comment (0xFelix) [reply pkg/virtctl/vnc/vnc.go:286]: Thinking about this, do we even need the if statement around the call to log?

Why not do it like this?

```
log.Log.V(4).Infof("Executing commandline: '%s %v'", vncBin, args)
```

**Review by 0xFelix — COMMENTED** [review]
> > There is initialization in virtctl/root.go in `Execute()` function, I think it may be useful if I called InitializeLogging in the init function of the `virtctl` package with `os.Args[0]` as program name
> 
> I think it need to be per module? So, on `root.go` logs, it'll appear the component just right but on `usbredir` it'll be empty.

I verified this locally, see component is empty right now:

```
felix@felix-p1 kubevirt/kubevirt (main) » virtctl vnc -v4 testvm                                                                                   1 ↵
{"component":"","level":"info","msg":"--proxy-only is set to false, listening on 127.0.0.1\n","pos":"vnc.go:116","timestamp":"2024-04-12T14:29:11.066744Z"}
{"component":"","level":"info","msg":"connection timeout: 1m0s","pos":"vnc.go:157","timestamp":"2024-04-12T14:29:11.067090Z"}
{"component":"","level":"info","msg":"Executing commandline: 'remote-viewer [vnc://127.0.0.1:34245 --debug]'","pos":"vnc.go:287","timestamp":"2024-04-12T14:29:11.067199Z"}
{"component":"","level":"info","msg":"VNC Client connected in 182.092051ms","pos":"vnc.go:170","timestamp":"2024-04-12T14:29:11.249183Z"}
```

Personally, I'd like if we could avoid `init` functions. I think a better place would be the `RunE` field of each command. This way we could place the logger initialization with the component in a common place for each module.

Also, could you please squash your commits with the fixes?

File: pkg/virtctl/vnc/vnc.go (hunk 7 above)
Comment (victortoso) [reply pkg/virtctl/vnc/vnc.go:286]: Here, not indeed. In line 314 though, using 9e5f2fc6422ed156282fff88643fb2c441a96764 can be helpful

Comment (Zaher1307) [reply pkg/virtctl/vnc/vnc.go:286]: I don't know but I think logging here depends on which verbosity level set before logging

**Comment by Zaher1307** [comment]
> > > There is initialization in virtctl/root.go in `Execute()` function, I think it may be useful if I called InitializeLogging in the init function of the `virtctl` package with `os.Args[0]` as program name
> > 
> > 
> > I think it need to be per module? So, on `root.go` logs, it'll appear the component just right but on `usbredir` it'll be empty.
> 
> I verified this locally, see component is empty right now:
> 
> ```
> felix@felix-p1 kubevirt/kubevirt (main) » virtctl vnc -v4 testvm                                                                                   1 ↵
> {"component":"","level":"info","msg":"--proxy-only is set to false, listening on 127.0.0.1\n","pos":"vnc.go:116","timestamp":"2024-04-12T14:29:11.066744Z"}
> {"component":"","level":"info","msg":"connection timeout: 1m0s","pos":"vnc.go:157","timestamp":"2024-04-12T14:29:11.067090Z"}
> {"component":"","level":"info","msg":"Executing commandline: 'remote-viewer [vnc://127.0.0.1:34245 --debug]'","pos":"vnc.go:287","timestamp":"2024-04-12T14:29:11.067199Z"}
> {"component":"","level":"info","msg":"VNC Client connected in 182.092051ms","pos":"vnc.go:170","timestamp":"2024-04-12T14:29:11.249183Z"}
> ```
> 
> Personally, I'd like if we could avoid `init` functions. I think a better place would be the `RunE` field of each command. This way we could place the logger initialization with the component in a common place for each module.
> 
> Also, could you please squash your commits with the fixes?

Ok, I will do it but what preferred program name that should be initialized in logging of each module?

**Comment by 0xFelix** [comment]
For a start I think the name of the running subcommand would be fine. (`virtctl usbredir` --> `"componenent": "usbredir"`). Also, can you put the initialization into the `RunE` field of the command definitions? E.g. [here](https://github.com/kubevirt/kubevirt/blob/01f3d4c4ec4d4920af7ed3c10d9acbc62fac8c9a/pkg/virtctl/usbredir/usbredir.go#L49), so it is in the same place for every command?

**Comment by 0xFelix** [comment]
@Zaher1307 Are you going to update other subcommands too?

**Comment by Zaher1307** [comment]
> @Zaher1307 Are you going to update other subcommands too?

Which other subcommands?

**Comment by Zaher1307** [comment]
> https://github.com/kubevirt/kubevirt/blob/d56984cb7160469949aff6899d087c5b94ef0cee/pkg/virtctl/root.go#L83-L113
> 
> You already touched some of those. I was thinking you could update them all.

logger is not associated with all of these commands, just the commands I added initialize logger function to

**Comment by 0xFelix** [comment]
@Zaher1307 Did not have the time to check, so the other commands do not use the logger, is that correct?

**Comment by Zaher1307** [comment]
> @Zaher1307 Did not have the time to check, so the other commands do not use the logger, is that correct?

yes it is correct

File: pkg/virtctl/portforward/portforwarder.go [hunk 11]
```diff
@@ -5,7 +5,8 @@ import (
 	"net"
 	"strings"
 
-	"github.com/golang/glog"
+	"kubevirt.io/client-go/kubecli"
```
[the remark this answers is not in the document]
Comment (Zaher1307) [reply pkg/virtctl/portforward/portforwarder.go]: I removed it

**Review by 0xFelix — COMMENTED** [review]
One nit, apart from that, thank you!

Also, before the merge, could you please squash your commits?

/approve

**Comment by Zaher1307** [comment]
> One nit, apart from that, thank you!
> 
> Also, before the merge, could you please squash your commits where it makes sense?
> 
> /approve

I have already merged main branch into this branch many times while development, that is why my local branch have many commits from main branch, and it is a tedious task to squash all of them.
you can simply use github option to squash and merge while merging.

**Comment by alicefr** [comment]
> I have already merged main branch into this branch many times while development, that is why my local branch have many commits from main branch, and it is a tedious task to squash all of them. you can simply use github option to squash and merge while merging.

@Zaher1307 please review our contributing[guide](https://github.com/kubernetes/community/blob/master/contributors/guide/contributing.m), in particularly the [best-practice section](https://github.com/kubernetes/community/blob/master/contributors/guide/contributing.md#best-practices). If you follow the guideline there it becomes easier for us and for you to get your changes merged.

**Comment by Zaher1307** [comment]
> @Zaher1307 The way we merge commits in this repo is to preserve all commits in a PR. IMO it should not be too hard for you to squash them to 3 commits. One commit for the switch to `kubevirt.io/client-go/log`, one commit for the removal of `AddGlogFlags` and one commit for the initialization of the logger. For a start: Running `git reset HEAD~11` from this branch should undo all of your commits while preserving the changes. Then you can add the relevant changes to the three new commits.

git reset HEAD~11 doesn't work properly with my case, I have merged main into my branch several time so the commits in my branch is not only what it committed.
check this:
image

**Comment by alicefr** [comment]
> > @Zaher1307 The way we merge commits in this repo is to preserve all commits in a PR. IMO it should not be too hard for you to squash them to 3 commits. One commit for the switch to `kubevirt.io/client-go/log`, one commit for the removal of `AddGlogFlags` and one commit for the initialization of the logger. For a start: Running `git reset HEAD~11` from this branch should undo all of your commits while preserving the changes. Then you can add the relevant changes to the three new commits.
> 
> git reset HEAD~11 doesn't work properly with my case, I have merged main into my branch several time so the commits in my branch is not only what it committed. check this: image

@Zaher1307 I would strongly encourage you to take some git tutorials. This is a standard flow for (almost) all github projects and a must have for open source. You need to have a clean git history, otherwise, there are no chances we can merged your changes.

For `git reset`, your files might be still staged and you need to unstaged the ones which don't belong to the current commit ([example](https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/git-unstage-file-all-index-commit-folder-add-delete)). Once you commits the file for the first, you can add the other and create the corresponding commit.
If you need to rebase on the `main` branch, what I usually do is go to the PR branch and run `git rebase main`. If you don't have conflicts, your changes should be put on top of the latest version of main.

**Pull request closed without merging**